# PopVar-Tutorial

## About PopVar

Implementing GS methodology, the R package PopVar(Mohammadi et al 2015) uses genotype and phenotype data of a set of parents to predict genetic variance and progeny means for all possible biparental crosses in a half-diallel mating design [(P*(P-1))/2, where P = number of parents] 

Here, you can find a basic script to run PopVar R package (Mohammadi et al 2015) and examples of input files. 
Input files contain subsetted data from SunGrains (http://www.sungrains.lsu.edu/) propietary data, therefore names of marker and genotypes have been changed. 

PopVar_Code.Rmd
Contains a basic PopVar R script 

MarkersData.csv
Contains an example of SNP dataset for PopVar

PhenoFile.csv
Contains an example of phenotype dataset for PopVar

GeneticMap.csv
Contains an example of genetic map file for PopVar

## References

Mohammadi, M., Tiede, T., and Smith, K.P. (2015). PopVar: A genome-wide procedure for predicting genetic variance and correlated response in biparental breeding populations. Crop Science 55, 2068-2077.
https://acsess.onlinelibrary.wiley.com/doi/full/10.2135/cropsci2015.01.0030
